app/server/lib/Client.ts

```diff
@@ -501,6 +501,10 @@ export class Client {
     const user = this._profile ? await this._fetchUser(dbManager) : dbManager.getAnonymousUser();
     this._user = user ? dbManager.makeFullUser(user) : undefined;
     this._firstLoginAt = user?.firstLoginAt || null;
+    if (this._user) {
```

Comment:
Is it important for you to also update `lastConnectionAt` when a user visits a non-document page? It's uncommon for a user to visit Grist without opening a document, but it could happen.

Reply:
Yes, I think it will be better, it's still an action in Grist.

Reply:
In that case, I think @fflorent's suggestion to modify `Authorizer.addRequestUser` is a reasonable approach.

Here's the relevant snippet of code modified to save the `lastConnectionAt` in the session object (if it differs from the current date). You'll need to tweak the types in a few other files, and possibly the types of the stored dates themselves. (I haven't run the code, so other changes may be necessary; this is just a rough starting point.)

```javascript
      ...
      // If we haven't computed a userId yet, check for one using an email address in the profile.
      // A user record will be created automatically for emails we've never seen before.
      if (profile && !mreq.userId) {
        const userOptions: UserOptions = {};
        if (profile?.loginMethod === 'Email + Password') {
          // Link the session authSubject, if present, to the user. This has no effect
          // if the user already has an authSubject set in the db.
          userOptions.authSubject = sessionUser.authSubject;
        }
        const today = moment().startOf('day');
        if (today !== moment(sessionUser.lastConnectionAt).startOf('day')) {
          userOptions.lastConnectionAt = today.toDate();
          const scopedSession = options.gristServer.getSessions().getOrCreateSessionFromRequest(req);
          await scopedSession.updateUser(req, {lastConnectionAt: today});
        }
        const user = await dbManager.getUserByLoginWithRetry(profile.email, {profile, userOptions});
        if (user) {
          mreq.user = user;
          mreq.userId = user.id;
          mreq.userIsAuthorized = true;
        }
      }
      ...
```

With something like the above, I don't think we'd need the changes to `Client.ts` anymore, right? (As opening a document should hit a home server endpoint, which will involve calling `Authorizer.addRequestUser`.)

Reply:
I think in fact it's a better place to modify `lastConnectionAt` but why modify the `scopedSession` ? The only place we need the information to be saved is in the table `users` so we want to use `dbManager.updateUser` and not `scopedSession.updateUser`. If the user not exists we don't need to save the information.

Reply:
That sounds even better.

(I thought having it in the session might come in handy with tracking doc worker activity later, but it looks like Authorizer already fetches and caches a copy of the user there too, so no need to stash anything in the session.)